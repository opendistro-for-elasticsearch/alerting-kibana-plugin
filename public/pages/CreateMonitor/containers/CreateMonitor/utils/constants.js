import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';

/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

export const BUCKET_COUNT = 5;

export const MATCH_ALL_QUERY = JSON.stringify({ size: 0, query: { match_all: {} } }, null, 4);

export const FORMIK_INITIAL_VALUES = {
  /* CONFIGURE MONITOR */
  name: '',
  disabled: false,
  frequency: 'interval',
  timezone: [],
  daily: 0,
  period: { interval: 1, unit: 'MINUTES' },
  weekly: { mon: false, tue: false, wed: false, thur: false, fri: false, sat: false, sun: false },
  monthly: { type: 'day', day: 1 },
  cronExpression: '0 */1 * * *',

  /* DEFINE MONITOR */
  searchType: 'graph',
  index: [],
  timeField: '',
  query: MATCH_ALL_QUERY,
  aggregationType: 'count',
  fieldName: [],
  overDocuments: 'all documents',
  groupedOverTop: 5,
  groupedOverFieldName: 'bytes',
  bucketValue: 1,
  bucketUnitOfTime: 'h', // m = minute, h = hour, d = day
  where: {
    fieldName: [],
    operator: OPERATORS_MAP.IS,
    fieldValue: '',
    fieldRangeStart: 0,
    fieldRangeEnd: 0,
  },
  detectorId: '',
};
