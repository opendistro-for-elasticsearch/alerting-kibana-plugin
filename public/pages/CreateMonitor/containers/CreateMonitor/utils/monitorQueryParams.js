import { SEARCH_TYPE } from '../../../../../utils/constants';

export const initializeFromQueryParams = queryParams => {
  return {
    searchType: queryParams.searchType || undefined,
    name: queryParams.name ? `${queryParams.name}-Monitor` : undefined,
    detectorId: queryParams.adId || undefined,
    period:
      queryParams.interval && queryParams.unit
        ? { interval: parseInt(queryParams.interval), unit: queryParams.unit }
        : undefined,
  };
};
