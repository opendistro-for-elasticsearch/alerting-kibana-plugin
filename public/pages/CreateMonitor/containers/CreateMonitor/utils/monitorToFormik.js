/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import { FORMIK_INITIAL_VALUES } from './constants';
import { SEARCH_TYPE, INPUTS_DETECTOR_ID } from '../../../../../utils/constants';

// Convert Monitor JSON to Formik values used in UI forms
export default function monitorToFormik(monitor) {
  const formikValues = _.cloneDeep(FORMIK_INITIAL_VALUES);
  if (!monitor) return formikValues;
  const {
    name,
    enabled,
    schedule: { cron: { expression: cronExpression = formikValues.cronExpression, timezone } = {} },
    inputs,
    ui_metadata: { schedule = {}, search = {} } = {},
  } = monitor;
  // Default searchType to query, because if there is no ui_metadata or search then it was created through API or overwritten by API
  // In that case we don't want to guess on the UI what selections a user made, so we will default to just showing the extraction query
  const { searchType = 'query', fieldName } = search;
  const isAD = searchType === SEARCH_TYPE.AD;

  return {
    /* INITIALIZE WITH DEFAULTS */
    ...formikValues,

    /* CONFIGURE MONITOR */
    name,
    disabled: !enabled,

    /* This will overwrite the fields in use by Monitor from ui_metadata */
    ...schedule,
    cronExpression,

    /* DEFINE MONITOR */
    ...search,
    searchType,
    fieldName: fieldName ? [{ label: fieldName }] : [],
    timezone: timezone ? [{ label: timezone }] : [],

    detectorId: isAD //&& _.get(inputs, '0.search.indices.0') === '.opendistro-anomaly-results*'
      ? _.get(inputs, INPUTS_DETECTOR_ID)
      : undefined,
    index: inputs[0].search.indices.map(index => ({ label: index })),
    query: JSON.stringify(inputs[0].search.query, null, 4),
  };
}
