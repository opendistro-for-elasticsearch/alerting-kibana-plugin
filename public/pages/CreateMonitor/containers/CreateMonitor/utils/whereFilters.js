/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import { OPERATORS_MAP } from '../../../components/MonitorExpressions/expressions/utils/constants';
import { DATA_TYPES } from '../../../../../utils/constants';

//TODO:: Breakdown to factory pattern for rules in-case we support multiple filters. This is just ease for the single one
export const OPERATORS_QUERY_MAP = {
  [OPERATORS_MAP.IS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? { match_phrase: { [label]: fieldValue } }
        : { term: { [label]: fieldValue } },
  },
  [OPERATORS_MAP.IS_NOT]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: { must_not: { match_phrase: { [label]: fieldValue } } },
          }
        : {
            bool: { must_not: { term: { [label]: fieldValue } } },
          },
  },
  [OPERATORS_MAP.IS_NULL]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({
      bool: { must_not: { exists: { field: fieldKey } } },
    }),
  },
  [OPERATORS_MAP.IS_NOT_NULL]: {
    query: ({ fieldName: [{ label: fieldKey }] }) => ({ exists: { field: fieldKey } }),
  },
  [OPERATORS_MAP.IS_GREATER]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_GREATER_EQUAL]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { gte: fieldValue } },
    }),
  },
  [OPERATORS_MAP.IS_LESS]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lt: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IS_LESS_EQUAL]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldValue }) => ({
      range: { [fieldKey]: { lte: fieldValue } },
    }),
  },

  [OPERATORS_MAP.IN_RANGE]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } },
    }),
  },
  [OPERATORS_MAP.NOT_IN_RANGE]: {
    query: ({ fieldName: [{ label: fieldKey }], fieldRangeStart, fieldRangeEnd }) => ({
      bool: { must_not: { range: { [fieldKey]: { gte: fieldRangeStart, lte: fieldRangeEnd } } } },
    }),
  },

  [OPERATORS_MAP.STARTS_WITH]: {
    query: ({ fieldName, fieldValue }) => ({ prefix: { [fieldName[0].label]: fieldValue } }),
  },

  [OPERATORS_MAP.ENDS_WITH]: {
    query: ({ fieldName, fieldValue }) => ({
      wildcard: { [fieldName[0].label]: `*${fieldValue}` },
    }),
  },
  [OPERATORS_MAP.CONTAINS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            query_string: { query: `*${fieldValue}*`, default_field: label },
          }
        : {
            wildcard: { [label]: `*${fieldValue}*` },
          },
  },
  [OPERATORS_MAP.NOT_CONTAINS]: {
    query: ({ fieldName: [{ label, type }], fieldValue }) =>
      type === DATA_TYPES.TEXT
        ? {
            bool: {
              must_not: { query_string: { query: `*${fieldValue}*`, default_field: label } },
            },
          }
        : {
            bool: { must_not: { wildcard: { [label]: `*${fieldValue}*` } } },
          },
  },
};
