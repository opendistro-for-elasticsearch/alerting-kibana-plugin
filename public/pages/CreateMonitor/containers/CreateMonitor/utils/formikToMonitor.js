/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import _ from 'lodash';
import moment from 'moment-timezone';
import { BUCKET_COUNT } from './constants';
import { SEARCH_TYPE } from '../../../../../utils/constants';
import { OPERATORS_QUERY_MAP } from './whereFilters';

export function formikToMonitor(values) {
  const uiSchedule = formikToUiSchedule(values);
  const schedule = buildSchedule(values.frequency, uiSchedule);
  const uiSearch = formikToUiSearch(values);
  return {
    name: values.name,
    type: 'monitor',
    enabled: !values.disabled,
    schedule,
    inputs: [formikToSearch(values)],
    triggers: [],
    ui_metadata: {
      schedule: uiSchedule,
      search: uiSearch,
    },
  };
}

export function formikToInputs(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  return [isAD ? formikToAd(values) : formikToSearch(values)];
}

export function formikToSearch(values) {
  const isAD = values.searchType === SEARCH_TYPE.AD;
  let query = isAD ? formikToAdQuery(values) : formikToQuery(values);
  const indices = isAD ? ['.opendistro-anomaly-results*'] : formikToIndices(values);

  return {
    search: {
      indices,
      query,
    },
  };
}

export function formikToAdQuery(values) {
  return {
    size: 1,
    sort: [{ anomaly_grade: 'desc' }, { confidence: 'desc' }],
    query: {
      bool: {
        filter: [
          {
            range: {
              data_end_time: {
                from: '{{period_end}}||-' + values.period.interval + 'm',
                to: '{{period_end}}',
                include_lower: true,
                include_upper: true,
              },
            },
          },
          {
            term: {
              detector_id: {
                value: values.detectorId,
              },
            },
          },
        ],
      },
    },
    aggregations: {
      max_anomaly_grade: {
        max: {
          field: 'anomaly_grade',
        },
      },
    },
  };
}

export function formikToAd(values) {
  return {
    anomaly_detector: {
      detector_id: values.detectorId,
    },
  };
}
export function formikToUiSearch(values) {
  const {
    searchType,
    aggregationType,
    timeField,
    fieldName: [{ label: fieldName = '' } = {}],
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  } = values;
  return {
    searchType,
    aggregationType,
    timeField,
    fieldName,
    overDocuments,
    groupedOverTop,
    groupedOverFieldName,
    bucketValue,
    bucketUnitOfTime,
    where,
  };
}

export function formikToIndices(values) {
  return values.index.map(({ label }) => label);
}

export function formikToQuery(values) {
  const isGraph = values.searchType === SEARCH_TYPE.GRAPH;
  return isGraph ? formikToGraphQuery(values) : formikToExtractionQuery(values);
}

export function formikToExtractionQuery(values) {
  return JSON.parse(values.query);
}

export function formikToGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const whenAggregation = formikToWhenAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          gte: `{{period_end}}||-${Math.round(bucketValue)}${bucketUnitOfTime}`,
          lte: '{{period_end}}',
          format: 'epoch_millis',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: whenAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiGraphQuery(values) {
  const { bucketValue, bucketUnitOfTime } = values;
  const overAggregation = formikToUiOverAggregation(values);
  const timeField = values.timeField;
  const filters = [
    {
      range: {
        [timeField]: {
          // default range window to [BUCKET_COUNT] * the date histogram interval
          gte: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          lte: 'now',
        },
      },
    },
  ];
  const whereClause = formikToWhereClause(values);
  if (whereClause) {
    filters.push({ ...whereClause });
  }
  return {
    size: 0,
    aggregations: overAggregation,
    query: {
      bool: {
        filter: filters,
      },
    },
  };
}

export function formikToUiOverAggregation(values) {
  const whenAggregation = formikToWhenAggregation(values);
  const { bucketValue, bucketUnitOfTime } = values;
  const timeField = values.timeField;

  return {
    over: {
      date_histogram: {
        field: timeField,
        interval: `${bucketValue}${bucketUnitOfTime}`,
        time_zone: moment.tz.guess(),
        min_doc_count: 0,
        extended_bounds: {
          min: `now-${bucketValue * BUCKET_COUNT}${bucketUnitOfTime}`,
          max: 'now',
        },
      },
      aggregations: whenAggregation,
    },
  };
}

export function formikToWhereClause({ where }) {
  if (where.fieldName.length > 0) {
    return OPERATORS_QUERY_MAP[where.operator].query(where);
  }
}

export function formikToWhenAggregation(values) {
  const {
    aggregationType,
    fieldName: [{ label: field } = {}],
  } = values;
  if (aggregationType === 'count' || !field) return {};
  return { when: { [aggregationType]: { field } } };
}

export function formikToUiSchedule(values) {
  return {
    timezone: _.get(values, 'timezone[0].label', null),
    frequency: values.frequency,
    period: values.period,
    daily: values.daily,
    weekly: values.weekly,
    monthly: values.monthly,
    cronExpression: values.cronExpression,
  };
}

export function buildSchedule(scheduleType, values) {
  const {
    period,
    daily,
    weekly,
    monthly: { type, day },
    cronExpression,
    timezone,
  } = values;
  switch (scheduleType) {
    case 'interval': {
      return { period };
    }
    case 'daily': {
      return { cron: { expression: `0 ${daily} * * *`, timezone } };
    }
    case 'weekly': {
      const daysOfWeek = Object.entries(weekly)
        .filter(([day, checked]) => checked)
        .map(([day]) => day.toUpperCase())
        .join(',');
      return { cron: { expression: `0 ${daily} * * ${daysOfWeek}`, timezone } };
    }
    case 'monthly': {
      let dayOfMonth = '?';
      if (type === 'day') {
        dayOfMonth = day;
      }
      return { cron: { expression: `0 ${daily} ${dayOfMonth} */1 *`, timezone } };
    }
    case 'cronExpression':
      return { cron: { expression: cronExpression, timezone } };
  }
}
