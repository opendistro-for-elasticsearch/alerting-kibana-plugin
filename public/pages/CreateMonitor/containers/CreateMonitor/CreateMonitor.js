/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React, { Component, Fragment } from 'react';
import _ from 'lodash';
import queryString from 'query-string';
import { Formik } from 'formik';
import {
  EuiSpacer,
  EuiTitle,
  EuiFlexGroup,
  EuiFlexItem,
  EuiButton,
  EuiButtonEmpty,
} from '@elastic/eui';

import ConfigureMonitor from '../ConfigureMonitor';
import DefineMonitor from '../DefineMonitor';
import { FORMIK_INITIAL_VALUES } from './utils/constants';
import monitorToFormik from './utils/monitorToFormik';
import { formikToMonitor } from './utils/formikToMonitor';
import { DefineSchedule } from '../DefineSchedule';
import { TRIGGER_ACTIONS, SEARCH_TYPE } from '../../../../utils/constants';
import { initializeFromQueryParams } from './utils/monitorQueryParams';

export default class CreateMonitor extends Component {
  static defaultProps = {
    edit: false,
    monitorToEdit: null,
    detectorId: null,
    updateMonitor: () => {},
  };

  constructor(props) {
    super(props);

    //pre-populate some of values if query params exists.
    let initialValues = _.mergeWith(
      {},
      _.cloneDeep(FORMIK_INITIAL_VALUES),
      initializeFromQueryParams(queryString.parse(this.props.location.search)),
      (initialValue, queryValue) => (_.isEmpty(queryValue) ? initialValue : queryValue)
    );

    if (this.props.edit && this.props.monitorToEdit) {
      initialValues = monitorToFormik(this.props.monitorToEdit);
    }

    this.state = { initialValues };

    this.onCancel = this.onCancel.bind(this);
    this.onCreate = this.onCreate.bind(this);
    this.onSubmit = this.onSubmit.bind(this);
    this.onUpdate = this.onUpdate.bind(this);
  }

  onCancel() {
    if (this.props.edit) this.props.history.goBack();
    else this.props.history.push('/monitors');
  }

  async onCreate(monitor, { setSubmitting, setErrors }) {
    const { httpClient } = this.props;
    try {
      const resp = await httpClient.post('../api/alerting/monitors', monitor);
      setSubmitting(false);
      const {
        data: {
          ok,
          resp: { _id },
        },
      } = resp;
      if (ok) {
        this.props.history.push(
          `/monitors/${_id}?action=${TRIGGER_ACTIONS.CREATE_TRIGGER}&success=true`
        );
      } else {
        console.log('Failed to create:', resp.data);
      }
    } catch (err) {
      console.error(err);
      setSubmitting(false);
      // TODO: setErrors
    }
  }

  async onUpdate(monitor, { setSubmitting, setErrors }) {
    const { updateMonitor } = this.props;
    // Since we are creating a monitor from formik values, we do not want to override the current existing triggers
    // delete the triggers key so when merging original monitor with updated monitor we are keeping the original triggers
    const updatedMonitor = _.cloneDeep(monitor);
    delete updatedMonitor.triggers;
    try {
      const resp = await updateMonitor(updatedMonitor);
      setSubmitting(false);
      const {
        data: { ok, id },
      } = resp;
      if (ok) {
        this.props.history.push(`/monitors/${id}`);
      } else {
        console.log('Failed to update:', resp.data);
      }
    } catch (err) {
      console.error(err);
      setSubmitting(false);
      // TODO: setErrors
    }
  }

  onSubmit(values, formikBag) {
    const { edit } = this.props;
    const monitor = formikToMonitor(values);
    if (edit) this.onUpdate(monitor, formikBag);
    else this.onCreate(monitor, formikBag);
  }

  render() {
    const { initialValues } = this.state;
    const { edit, httpClient, monitorToEdit } = this.props;
    return (
      <div style={{ padding: '25px 50px' }}>
        <Formik
          initialValues={initialValues}
          onSubmit={this.onSubmit}
          validateOnChange={false}
          render={({ values, errors, handleSubmit, isSubmitting }) => (
            <Fragment>
              <EuiTitle size="l">
                <h1>{edit ? 'Edit' : 'Create'} monitor</h1>
              </EuiTitle>
              <EuiSpacer />
              <ConfigureMonitor httpClient={httpClient} monitorToEdit={monitorToEdit} />
              <EuiSpacer />
              <DefineMonitor
                values={values}
                errors={errors}
                httpClient={httpClient}
                detectorId={this.props.detectorId}
              />
              <Fragment>
                <EuiSpacer />
                <DefineSchedule isAd={values.searchType === SEARCH_TYPE.AD} />
              </Fragment>
              <EuiSpacer />
              <EuiSpacer />
              <EuiFlexGroup alignItems="center" justifyContent="flexEnd">
                <EuiFlexItem grow={false}>
                  <EuiButtonEmpty onClick={this.onCancel}>Cancel</EuiButtonEmpty>
                </EuiFlexItem>
                <EuiFlexItem grow={false}>
                  <EuiButton fill onClick={handleSubmit} isLoading={isSubmitting}>
                    {edit ? 'Update' : 'Create'}
                  </EuiButton>
                </EuiFlexItem>
              </EuiFlexGroup>
            </Fragment>
          )}
        />
      </div>
    );
  }
}
