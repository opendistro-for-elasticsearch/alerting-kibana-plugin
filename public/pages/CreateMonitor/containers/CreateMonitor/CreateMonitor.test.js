/*
 *   Copyright 2019 Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *   Licensed under the Apache License, Version 2.0 (the "License").
 *   You may not use this file except in compliance with the License.
 *   A copy of the License is located at
 *
 *       http://www.apache.org/licenses/LICENSE-2.0
 *
 *   or in the "license" file accompanying this file. This file is distributed
 *   on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
 *   express or implied. See the License for the specific language governing
 *   permissions and limitations under the License.
 */

import React from 'react';
import { shallow } from 'enzyme';

import CreateMonitor from './CreateMonitor';
import { historyMock, httpClientMock } from '../../../../../test/mocks';
import { FORMIK_INITIAL_VALUES } from './utils/constants';
import AlertingFakes from '../../../../../test/utils/helpers';
import { SEARCH_TYPE } from '../../../../utils/constants';

const alertingFakes = new AlertingFakes('CreateMonitor random seed');

const setFlyout = jest.fn();
const updateMonitor = jest.fn().mockResolvedValue({ data: { ok: true, id: 'monitor_id' } });
const formikBag = {
  setSubmitting: jest.fn(),
  setErrors: jest.fn(),
};
const match = {
  isExact: true,
  params: {},
  path: '/create-monitor',
  url: '/create-monitor',
};
const location = {
  hash: '',
  pathname: '/create-monitor',
  search: '',
  state: undefined,
};
beforeEach(() => {
  jest.clearAllMocks();
});

describe('CreateMonitor', () => {
  test('renders', () => {
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        history={historyMock}
        setFlyout={setFlyout}
        match={match}
        location={location}
      />
    );
    expect(wrapper).toMatchSnapshot();
  });

  test('uses AD query Params as initialValues when parameter exists', () => {
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        history={historyMock}
        setFlyout={setFlyout}
        match={match}
        location={{ ...location, search: 'searchType=ad&adId=randomId&name=Sample' }}
      />
    );
    expect(wrapper.instance().state.initialValues.name).toBe('Sample-Monitor');
    expect(wrapper.instance().state.initialValues.searchType).toBe(SEARCH_TYPE.AD);
    expect(wrapper.instance().state.initialValues.detectorId).toBe('randomId');
  });

  test('uses monitorToEdit as initialValues when editing', () => {
    const monitor = alertingFakes.randomMonitor();
    const wrapper = shallow(
      <CreateMonitor
        httpClient={httpClientMock}
        edit={true}
        history={historyMock}
        setFlyout={setFlyout}
        updateMonitor={updateMonitor}
        monitorToEdit={monitor}
        match={match}
        location={location}
      />
    );
    expect(wrapper.instance().state.initialValues.name).toBe(monitor.name);
  });

  describe('onSubmit', () => {
    test('calls only onUpdate when editing', () => {
      const onCreate = jest.spyOn(CreateMonitor.prototype, 'onCreate');
      const onUpdate = jest.spyOn(CreateMonitor.prototype, 'onUpdate');
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(onUpdate).toHaveBeenCalledTimes(1);
      expect(onCreate).not.toHaveBeenCalled();
    });

    test('calls only onCreate when creating', () => {
      const onCreate = jest.spyOn(CreateMonitor.prototype, 'onCreate');
      const onUpdate = jest.spyOn(CreateMonitor.prototype, 'onUpdate');
      httpClientMock.post.mockResolvedValue({ data: { ok: true, resp: { _id: 'monitor_id' } } });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onSubmit(FORMIK_INITIAL_VALUES, formikBag);
      expect(onCreate).toHaveBeenCalledTimes(1);
      expect(onUpdate).not.toHaveBeenCalled();
    });
  });

  describe('onCancel', () => {
    test('calls history.goBack if editing', () => {
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onCancel();
      expect(historyMock.goBack).toHaveBeenCalledTimes(1);
      expect(historyMock.push).not.toHaveBeenCalled();
    });

    test('calls history.push when creating', () => {
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onCancel();
      expect(historyMock.push).toHaveBeenCalledTimes(1);
      expect(historyMock.goBack).not.toHaveBeenCalled();
    });
  });

  describe('onUpdate', () => {
    test('calls updateMonitor with monitor with no triggers key', () => {
      const monitor = alertingFakes.randomMonitor();
      const trigger = alertingFakes.randomTrigger();
      monitor.triggers = [trigger];
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onUpdate(monitor, formikBag);
      expect(updateMonitor).toHaveBeenCalledTimes(1);
      // The updatedMonitor that is passed in should NOT have the trigger key
      const updatedMonitor = { ...monitor };
      delete updatedMonitor.triggers;
      expect(updateMonitor).toHaveBeenCalledWith(updatedMonitor);
    });

    test('logs error when updateMonitor rejects', async () => {
      const error = jest.spyOn(global.console, 'error');
      updateMonitor.mockRejectedValue(new Error('updateMonitor error'));
      const monitor = alertingFakes.randomMonitor();
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      await wrapper.instance().onUpdate(monitor, formikBag);
      expect(error).toHaveBeenCalled();
    });

    test('logs resp.data when ok:false', async () => {
      const log = jest.spyOn(global.console, 'log');
      updateMonitor.mockResolvedValue({ data: { ok: false, resp: 'test' } });
      const monitor = alertingFakes.randomMonitor();
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          edit={true}
          history={historyMock}
          setFlyout={setFlyout}
          updateMonitor={updateMonitor}
          monitorToEdit={null}
          match={match}
          location={location}
        />
      );
      await wrapper.instance().onUpdate(monitor, formikBag);
      expect(log).toHaveBeenCalled();
      expect(log).toHaveBeenCalledWith('Failed to update:', { ok: false, resp: 'test' });
    });
  });

  describe('onCreate', () => {
    test('calls post with monitor', () => {
      const monitor = alertingFakes.randomMonitor();
      httpClientMock.post.mockResolvedValue({ data: { ok: true, resp: { _id: 'id' } } });
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      wrapper.instance().onCreate(monitor, formikBag);
      expect(httpClientMock.post).toHaveBeenCalledTimes(1);
      expect(httpClientMock.post).toHaveBeenCalledWith('../api/alerting/monitors', monitor);
    });

    test('logs error when updateMonitor rejects', async () => {
      const error = jest.spyOn(global.console, 'error');
      httpClientMock.post.mockRejectedValue(new Error('onCreate error'));
      const monitor = alertingFakes.randomMonitor();
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      await wrapper.instance().onCreate(monitor, formikBag);
      expect(error).toHaveBeenCalled();
    });

    test('logs resp.data when ok:false', async () => {
      const log = jest.spyOn(global.console, 'log');
      httpClientMock.post.mockResolvedValue({ data: { ok: false, resp: 'test' } });
      const monitor = alertingFakes.randomMonitor();
      const wrapper = shallow(
        <CreateMonitor
          httpClient={httpClientMock}
          history={historyMock}
          setFlyout={setFlyout}
          match={match}
          location={location}
        />
      );
      await wrapper.instance().onCreate(monitor, formikBag);
      expect(log).toHaveBeenCalled();
      expect(log).toHaveBeenCalledWith('Failed to create:', { ok: false, resp: 'test' });
    });
  });
});
