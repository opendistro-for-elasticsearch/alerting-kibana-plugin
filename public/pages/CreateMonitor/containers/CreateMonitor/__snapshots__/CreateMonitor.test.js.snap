// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateMonitor renders 1`] = `
<div
  style={
    Object {
      "padding": "25px 50px",
    }
  }
>
  <Formik
    enableReinitialize={false}
    initialValues={
      Object {
        "aggregationType": "count",
        "bucketUnitOfTime": "h",
        "bucketValue": 1,
        "cronExpression": "0 */1 * * *",
        "daily": 0,
        "detectorId": "",
        "disabled": false,
        "fieldName": Array [],
        "frequency": "interval",
        "groupedOverFieldName": "bytes",
        "groupedOverTop": 5,
        "index": Array [],
        "monthly": Object {
          "day": 1,
          "type": "day",
        },
        "name": "",
        "overDocuments": "all documents",
        "period": Object {
          "interval": 1,
          "unit": "MINUTES",
        },
        "query": "{
    \\"size\\": 0,
    \\"query\\": {
        \\"match_all\\": {}
    }
}",
        "searchType": "graph",
        "timeField": "",
        "timezone": Array [],
        "weekly": Object {
          "fri": false,
          "mon": false,
          "sat": false,
          "sun": false,
          "thur": false,
          "tue": false,
          "wed": false,
        },
        "where": Object {
          "fieldName": Array [],
          "fieldRangeEnd": 0,
          "fieldRangeStart": 0,
          "fieldValue": "",
          "operator": "is",
        },
      }
    }
    isInitialValid={false}
    onSubmit={[Function]}
    render={[Function]}
    validateOnBlur={true}
    validateOnChange={false}
  />
</div>
`;
